src/internal/tokencache/default_storage_manager.go

```diff
@@ -96,37 +96,38 @@ func (m *defaultStorageManager) ReadRefreshToken(
 	defer lock.RUnlock()
 	if familyID != "" {
```

Comment:
Interesting way of handling FOCI. I think it works. 

In .NET the algorithm we went with is: 

Simple algo: 

1. Try to find an FRT (i.e. ignore clientID filter)
2. If found, try to redeem the FRT
3. If AAD returns a token, hurray! Client is part of the family and we got a token.
3.1. If however AAD returns error code "client_mismatch" ( I think you need to ask AAD to add MSAL.GO on a special list for this to be returned) or "invalid_grant", then you know the client ID making the request is not part of the family

4. Assuming that FRT was either not found or it could not be redeemed, try to find the normal RT (filter by clientID)

Reply:
More complex algo involves remembering in the token cache which ClientIDs are part of the family. MSAL doesn't know this from the start, so assume a client is part of the family if you don't know otherwise.

Reply:
Is it worth writing an issue for?

Reply:
Is there an issue? I see msal.go already does the app metadata trick. I was trying to give some context around FOCI, not suggesting there is a bug.

Reply:
The implementation seems a little different, looks like this one doesn't handle client_mismatch / invalid_grant and I don't think failover from FRT to RT

Reply:
That's probably fine. I do wonder what happens if an app which is part of the family is removed from the family? Will MSAL go into an infinite loop or will the user just have to re-login interactively ? (second variant is acceptable imo )

Reply:
leaving the family is supported, as far as I recall

---
In reply to: [506577218](https://github.com/AzureAD/microsoft-authentication-library-for-go/pull/50#discussion_r506577218) [](ancestors = 506577218)

Reply:
If you leave the family, then on refreshing the FRT you get "client_mismatch" or "invalid_grant" from EVO, which msal MUST interpret as a `UIRequiredException`.

After interactive login, the app metadata will reflect the new state of world, where the app is no longer in the family. 

Should be fairly easy to unit test.